python/mxnet/contrib/quantization.py

```diff
@@ -303,32 +305,34 @@ def _get_optimal_threshold(arr, num_bins=8001, num_quantized_bins=255):
         right_outlier_count = np.sum(hist[p_bin_idx_stop:])
         p[-1] += right_outlier_count
         # is_nonzeros[k] indicates whether hist[k] is nonzero
-        is_nonzeros = (sliced_nd_hist != 0).astype(np.int32)
+        is_nonzeros = (p != 0).astype(np.int32)
 
         # calculate how many bins should be merged to generate quantized distribution q
-        num_merged_bins = p.size // num_quantized_bins
+        num_merged_bins = sliced_nd_hist.size // num_quantized_bins
         # merge hist into num_quantized_bins bins
         for j in range(num_quantized_bins):
             start = j * num_merged_bins
             stop = start + num_merged_bins
             quantized_bins[j] = sliced_nd_hist[start:stop].sum()
         quantized_bins[-1] += sliced_nd_hist[num_quantized_bins * num_merged_bins:].sum()
         # expand quantized_bins into p.size bins
-        q = np.zeros(p.size, dtype=np.float32)
+        q = np.zeros(sliced_nd_hist.size, dtype=np.float32)
         for j in range(num_quantized_bins):
             start = j * num_merged_bins
             if j == num_quantized_bins - 1:
-                stop = -1
+                stop = len(is_nonzeros)
             else:
                 stop = start + num_merged_bins
             norm = is_nonzeros[start:stop].sum()
             if norm != 0:
-                q[start:stop] = float(quantized_bins[j]) / float(norm)
-        q[sliced_nd_hist == 0] = 0
+                q[start:stop] = float(quantized_bins[j]) / float(num_quantized_bins)
```

Comment:
Originally this was `float(norm)`, and that is not appropriate. Suppose you have the distribution:

```
[0, 0, ... , 0, 1]
```

If the `num_quantized_bins` is `3`, then you theoretically should get:

```
[0, 0, ... , 1/3, 1/3, 1/3]
```

instead of:

```
[0, 0, ..., 1, 1, 1]
```

To make this more clear, suppose your original dist is:

```
[0, 0, 0, ... , 1/3, 1/3, 1/3]
```

This should be equivalent after quantization as the first distribution, but it isn't. Under the rules, you would get the same array back, where you have `[..., 1/3, 1/3 ,1/3]`, and the first distribution would give you `[..., 1, 1, 1]`, off by the multiplier.